flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -473,25 +519,41 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
             }
         }
 
-        if ((flags & NO_SUBCLASS) != 0) {
+        if (isRecord) {
             try {
+                JavaRecordBuilderFactory<T>.JavaRecordBuilder builder = recordHelper.newBuilder();
                 for (int i = 0; i < numFields; i++) {
                     boolean isNull = source.readBoolean();
 
                     if (fields[i] != null) {
                         if (isNull) {
-                            fields[i].set(reuse, null);
+                            builder.setField(i, null);
                         } else {
-                            Object field;
+                            Object reuseField = reuse == null ? null : fields[i].get(reuse);
```

Comment:
`reuse` will never be `null` here. We should reiterate over the code here. We create quite a bit of redundant code. The for loop and the exception handling are the same. The only difference is how the field values are set. We could use some consumer to out-source that and move the rest of the code into its own method. WDYT?